File: C++/Stack/Implementation/Linked List/Stack.cpp
//
// Created by vivek on 01-01-2023.
//

#include <iostream>
#include "Stack.h"

Stack::Stack() : top(nullptr) {}            // Constructor

void Stack::push(int x) {           // Push operation: insert element into stack
    if (!top)
        top = new Node(x);
    else {
        Node *t = new Node(x);
        t->next = top;
        top = t;
    }
}

void Stack::pop() {         // Pop operation: remove element from stack
    if (!top)
        std::cerr << "Stack underflow!\n";
    else {
        if (!top->next) {
            delete top;
            top = nullptr;
        } else {
            Node *t = top;
            top = top->next;
            delete t;
        }
    }
}

std::ostream &operator<<(std::ostream &os, const Stack &stack) {            // Printing the stack
    Node *t = stack.top;
    while (t) {
        os << t->data << std::endl;
        t = t->next;
    }
    os << std::endl;
    return os;
}

int Stack::peek() {         // Getting the top element of the stack
    return top ? top->data : -1;
}

Stack::~Stack() {           // Destructor to free up heap memory used by linked list
    Node *t = this->top;
    while (t) {
        top = top->next;
        delete t;
        t = top;
    }
}

File: Java/Stack/src/stack/implementation/linkedlist/Stack.java
package stack.implementation.linkedlist;

public class Stack {

    private Node top;

    public Stack() {
        this.top = null;
    }

    private void push(int x) {
        if (top == null) top = new Node(x, null);
        else top = new Node(x, top);
    }

    private void pop() {
        if (top == null) System.out.println("Stack underflow!");
        else top = (top.next == null) ? null : top.next;
    }

    private int peek() {
        return top != null ? top.data : -1;
    }

    @Override
    public String toString() {
        Node t = this.top;
        StringBuilder s = new StringBuilder();
        while (t != null) {
            s.append(t.data);
            if (t.next != null) s.append(", ");
            t = t.next;
        }
        s.append("\n");
        return s.toString();
    }

    private record Node(int data, Stack.Node next) {
    }

    public static void main(String[] args) {
        Stack stk = new Stack();
        stk.push(1);
        stk.push(2);
        stk.push(3);
        stk.pop();
        stk.pop();
        System.out.println(stk);
    }
}
